src/ime.cpp

```diff
@@ -0,0 +1,145 @@
+        pImmGetContext_t pImmGetContext;
+        pImmGetOpenStatus_t pImmGetOpenStatus;
+        pImmSetOpenStatus_t pImmSetOpenStatus;
+        pImmReleaseContext_t pImmReleaseContext;
+
+        template<typename T>
+        static T fun_ptr_cast( FARPROC p ) {
+            // workaround function cast warning
+            return reinterpret_cast<T>( reinterpret_cast<void *>( p ) );
+        }
+    public:
+        imm_wrapper() {
+            // Check if East Asian support is available
+            hImm = LoadLibrary( "imm32.dll" );
+            if( hImm ) {
+                pImmGetContext = fun_ptr_cast<pImmGetContext_t>(
+                                     GetProcAddress( hImm, "ImmGetContext" ) );
+                pImmGetOpenStatus = fun_ptr_cast<pImmGetOpenStatus_t>(
+                                        GetProcAddress( hImm, "ImmGetOpenStatus" ) );
+                pImmSetOpenStatus = fun_ptr_cast<pImmSetOpenStatus_t>(
+                                        GetProcAddress( hImm, "ImmSetOpenStatus" ) );
+                pImmReleaseContext = fun_ptr_cast<pImmReleaseContext_t>(
+                                         GetProcAddress( hImm, "ImmReleaseContext" ) );
+                if( !pImmGetContext || !pImmGetOpenStatus ||
+                    !pImmSetOpenStatus || !pImmReleaseContext ) {
+
+                    FreeLibrary( hImm );
+                    hImm = nullptr;
+                    pImmGetContext = nullptr;
+                    pImmGetOpenStatus = nullptr;
+                    pImmSetOpenStatus = nullptr;
+                    pImmReleaseContext = nullptr;
+                }
+            }
+        }
+
+        ~imm_wrapper() {
+            if( hImm ) {
+                FreeLibrary( hImm );
+            }
+        }
+
+        bool ime_enabled() {
+            if( hImm ) {
+                const HWND hwnd = getWindowHandle();
+                const HIMC himc = pImmGetContext( hwnd );
+                bool enabled = pImmGetOpenStatus( himc );
+                pImmReleaseContext( hwnd, himc );
+                return enabled;
+            }
+            return false;
+        }
+
+        void enable_ime() {
+            if( hImm ) {
+                const HWND hwnd = getWindowHandle();
+                const HIMC himc = pImmGetContext( hwnd );
+                pImmSetOpenStatus( himc, TRUE );
+                pImmReleaseContext( hwnd, himc );
+            }
+        }
+
+        void disable_ime() {
+            if( hImm ) {
+                const HWND hwnd = getWindowHandle();
+                const HIMC himc = pImmGetContext( hwnd );
+                pImmSetOpenStatus( himc, FALSE );
+                pImmReleaseContext( hwnd, himc );
+            }
+        }
+};
+
+static imm_wrapper imm;
+#endif
+
+static bool ime_enabled()
+{
+#if defined( __ANDROID__ )
+    return false; // always call disable_ime() (i.e. do nothing) on return
+#elif defined( _WIN32 )
+    return imm.ime_enabled();
+#endif
+    // TODO: other platforms?
```

Comment:
At the very least add a `return false;` here so the compiler won't complain abount missing return value.